# Los números notables (optativa)



$0$ :  elemento neutro de la suma.

$1$ : elemento neutro de la multiplicación.

$i$ : raíz cuadrada de -1.

$\pi$ : la razón entre el perímetro de la circunferencia y el diámetro.

$e$ : es el límite de la sucesión $\left(1 + \dfrac{1}{n}\right)^n$. Es decir,
$$
e = \lim_{n \to \infty} \left(1 + \dfrac{1}{n}\right)^n
$$

Veremos en lo que sigue formas de calcular $\pi$ y $e$.

##1. Cálculo de $\pi$

Veremos diferentes métodos para calcular $\pi$ que nos servirán para ilustrar diferentes estrategias de cálculo, muy distintas entre si.

El primer método no es ni fue utilizado para calcular $\pi$, pero ilustra un método probabilistico de cálculo,  usando *Montecarlo.*

El segundo método se inspira en un método usado por Arquímedes y en occidente hasta mediados del segundo milenio, el *método poligonal de Arquímedes*. Sin embargo  difiere en el método original en muchos aspectos. La similitud es, y ahí termina, en que se usan polígonos inscriptos y circunscriptos a una circunferencia para el cálculo de $\pi$.

El tercer método es el que desarrolló Newton a partir del desarrollo de $(a + b)^n$ (el *binomio de Newton*).

El cuarto método es un método relativamente moderno que calcula en forma eficiente aproximaciones de $\pi$

In [ ]:
# necesitamos importar estas bibliotecas (ya veremos por qué)
import random
import math

### Calcular $\pi$ contando las gotas de la lluvia.

El método que usaremos es *Montecarlo* y es muy intuitivo.

Se "dibuja" (en forma imaginaria) un cuadrado de $1 \times 1$ y se "tiran al azar" puntos $(x,y)$  en el cuadrado, es decir $0 \le  x, y \le 1$.

La distancia del origen a  $(x,y)$  es  $\sqrt{x^2 + y^2}$. Los puntos $(x,y)$ dentro de la circunferencia de radio $1$ son
$$
D^1 = \left\{(x,y):\sqrt{x^2 + y^2} \le 1\right\}
$$

Es decir, $(x,y)$ pertenece al círculo de radio $1$ sii $x^2 + y^2 \le 1$.

Ahora bien, la superficie del círculode radio 1 es $\pi$ por radiom al cuadrado. En  este caso  el radio  es $1/2$, es decir la superficie del círculo de radio $1$  es $\frac\pi4$. Como la superficie del cuadrado de lado $1$ es $1$, si tiramos $n$ puntos al azar en un cuadrado aproximadamente $n \cdot \frac\pi4$ caerán dentro del círculo.  Es decir si $k$ son los puntos que han caido en el cículo:
$$
k \sim \pi\cdot \frac{1}{4}\cdot  n.
$$
Haciendo pasaje de término,
$$
\pi \sim \frac{4k}{n}
$$

Implementemos un simulador de tirar puntos en el cuadrado que detecte los puntos que caen dentro del círculo:

In [ ]:
def calcular_pi_prob(n: int ) -> float:
    # calcula pi con n puntos
    puntos_en_circulo = 0
    for _ in range(n):
        x, y = random.uniform(0, 1), random.uniform(0, 1)
        if x**2 + y**2 <= 1:
            puntos_en_circulo = puntos_en_circulo + 1
    return 4 * puntos_en_circulo / n

print(calcular_pi_prob(100))
print(calcular_pi_prob(1000))
print(calcular_pi_prob(10**4))
print(calcular_pi_prob(10**5))
print(calcular_pi_prob(10**6))
print(calcular_pi_prob(10**7))

3.16
3.196
3.16
3.14156
3.143504
3.1416868


como verán la precisión no es muy buena, incluso tirando muchos puntos. Pero al menos nos da una idea del valor de $\pi$.

### Calcular $\pi$ con polígonos regulares

Cuando la longitud de la circunferencia se divide por la longitud del diámetro, el valor obtenido siempre se puede redondear a $3.14$ un valor cercano a $\pi$.

Una de las técnicas que se pueden utilizar para encontrar una aproximación más precisa al valor de $\pi$ fue desarrollada por Arquímedes. El método utilizado por él consiste en calcular una aproximación del valor $\pi$ mediante polígonos con perímetros muy similares a los de una circunferencia.





#### Versión simplificada del método poligonal de Arquímedes

Primero, dibujamos una circunferencia de radio $r$ y se construyen un cuadrado inscrito y otro cuadrado circunscrito respecto a esta circunferencia. Calculamos el perímetro de ambos cuadrados y los denotamos $p$ y $P$, al perímetro del inscrito y del circunscrito,  respectivamente. Si $C$ es el perímetro de la circunferencia,  entonces valen las siguientes desigualdades:
$$
p < C < P.
$$

El perímetro de la circunferencia de radio $r$ es  $C =2\pi r$, luego:
$$
p < C < P\quad \Rightarrow \quad \frac{p}{2r} < \pi < \frac{P}{2r}.
$$
Por lo tanto, $\pi$ es un número que se encuentra entre los perímetros del cuadrado más pequeño y el cuadrado más grande, ambos divididos por diámetro.



Esto se puede generalizar a polígonos regulares de $n$-lados: si consideramos $r=1$, $p_n$ el perímetro del polígono regular de $n$ lados inscrito en  la circunferencia de radio 1,  $P_n$ el perímetro del polígono regular de $n$ lados circunscrito en  la circunferencia de radio 1 y $C_1 = 2\pi$ el perímetro de la circunferencia de radio 1,  entonces
\begin{equation}
p_n < 2\pi < P_n\quad \Rightarrow  \quad \frac{p_n}{2} < \pi < \frac{P_n}{2}. \tag{1}
\end{equation}

Ahora bien $p_n = n \cdot b_n$ donde $b_n$  es la longitud de un lado del polígono. Un poco de trigonometría nos dice que $b_n = 2\sin({\frac{360}{n}/2}) = 2\sin(\frac{180}{n})$. Luego $p_n = 2n \sin(\frac{180}{n})$.

De  forma análoga, con un poco de trigonometría,  $P_n = n \cdot B_n$ donde $B_n$  es la longitud del lado y $B_n = 2\tan({\frac{360}{n}/2}) = 2\tan(\frac{180}{n})$. Luego $P_n = 2n \tan(\frac{180}{n})$.

Concluyendo:  reemplazando en la fórmula (1) obtenemos
$$
p_n < C_1 < P_n\quad => \quad n \sin(\frac{180}{n}) < \pi < n \tan(\frac{180}{n}).
$$

In [ ]:
def  calcular_pi_poligono_ins(n: int) -> float:
    # pre: n >=3
    # post: calcula pi haciendo el polígono regular de n-lados inscripto en una circunferencia de radio 1
    angulo = 2 * math.pi / n
    base = 2 * math.sin(angulo / 2)
    perimetro = n * base # parecido a  2 * PI * radio = 2 * PI
    return perimetro / 2

print(calcular_pi_poligono_ins(10000))

def  calcular_pi_poligono_cir(n: int) -> float:
    # pre: n >=3
    # post: calcula pi haciendo el polígono regular de n-lados circunscripto a una circunferencia de radio 1
    angulo = 2 * math.pi / n
    base = 2 * math.tan(angulo / 2)
    perimetro = n * base # parecido  2 * PI * radio = 2 * PI
    return perimetro / 2

print(calcular_pi_poligono_cir(10000))

print(calcular_pi_poligono_cir(10000) - calcular_pi_poligono_ins(10000))



3.141592601912665
3.141592756944053
1.550313877274334e-07


Este método tiene de interesante que nos da una estimación del error que se está cometiendo. Por ejemplo,  como vemos arriba,  con polígonos de 10000 lados sabemos que obtenemos $\pi$  hasta el 6º decimal.

El método de Arquímedes se basaba en que sabiendo los lados de un triángulo de $n$ lados se podía calcular los lados de un triángulo de $2n$ lados, pero esto involucraba unas cuentas muy trabajosas. Con este método calculó aproximaciones de $\pi$ con polígonos de $3$, $6$, $12$, $24$, $48$ y $96$ lados. Este último cálculo nos da que:

In [ ]:
print(calcular_pi_poligono_cir(96))
print(calcular_pi_poligono_ins(96))
print(calcular_pi_poligono_cir(96) - calcular_pi_poligono_ins(96))

3.1427145996453683
3.1410319508905093
0.0016826487548589064


y por lo tanto
$$
3.1410319508905093 < \pi <  3.1427145996453683.
$$
y la precisión es de 2 decimales.

### Calcular $\pi$ con el método de Newton

La fórmula que obutvo Newton se sigue de un  caso particular de la derivada de la la función arcoseno. Está fuera del alcance de este curso la justificación y demostración de la formula y  solamente vermos  su implementación.

Para mas datos ver:

-  http://www.pi314.net/eng/newton.php , o

- https://www.youtube.com/watch?v=9EJqxZqf63I


La fórmula es la siguiente:

$$
\pi =  24\left(\frac{\sqrt[2]{3}}{32}- \sum_{k=0}^\infty \frac{ (2k)!}{2^{4k+2}(k!)^2  (2k - 1)  (2k + 3))}\right)
 $$




 Esto significa que $\pi$ se puede aproximar por la suma "hasta $n$". Es decir  dado $n \ge 0$,
 $$
 \pi/24 \sim  \frac{\sqrt[2]{3}}{32}- \sum_{k=0}^n \frac{ (2k)!}{2^{4k+2}(k!)^2  (2k - 1)  (2k + 3))}
 $$
Dicho de otra forma, si  
$$a_k = \frac{ (2k)!}{2^{4k+2}(k!)^2  (2k - 1)  (2k + 3))}
$$
entonces
$$
\pi/24 \sim  \frac{\sqrt[2]{3}}{32}- \sum_{k=0}^n a_k
$$
o
$$
\pi/24 \sim  \frac{\sqrt[2]{3}}{32}- a_0 - a_1 - a_2 - \cdots -a_n.
$$

La implementación es sencilla utilizando la función `math.factorial()`, el factorial.

In [ ]:
def calcular_pi_Newton(n: int ) -> float:
    # ver: http://www.pi314.net/eng/newton.php
    # calcula  24*(3**0.5 /32 - \sum_{k=0}^n  0.5**(4*k+2) * (2*k)!/ (k!**2 * (2*k - 1)*(2*k + 3)))
    pi24 = 3**0.5 / 32
    for k in range(n+1):
        pi24 = pi24 -  math.factorial(2*k) / (2**(4*k+2) * math.factorial(k)**2 * (2*k - 1) * (2*k + 3))
    return 24 * pi24

print(calcular_pi_Newton(50)) # 30 decimales correctos

3.1415926535897936


### Cálculo de $\pi$ moderno (una opción)

Ver https://es.wikipedia.org/wiki/F%C3%B3rmula_de_Bailey-Borwein-Plouffe

La fórmula Bailey-Borwein-Plouffe (fórmula BBP) es una fórmula para $\pi$. Fue descubierta en 1995 por Simon Plouffe y lleva el nombre de los autores del artículo en el que fue publicado, David H. Bailey, Peter Borwein y Plouffe. La fórmula es
$$
\pi = \sum_{k = 0}^\infty \frac{1}{16^k}
\left( \frac{4}{8k + 1} - \frac{2}{8k + 4} - \frac{1}{8k + 5} - \frac{1}{8k + 6}\right).
$$

Esta fómula permite, aunque no de manera obvia, calular el  $n$-ésimo dígito de $\pi$ escrito en base $16$ sin calcular los dígitos anteriores.

Lo  que no es difícl de ver, basándose en la teoría de escribir un número en una base dada, que el error de las sumatoria hasta $n$  difiere con $\pi$ en menos de $1/16^n$. Es decir
$$
0 < \pi -  \sum_{k = 0}^n \frac{1}{16^k}
\left( \frac{4}{8k + 1} - \frac{2}{8k + 4} - \frac{1}{8k + 5} - \frac{1}{8k + 6}\right) < \frac{1}{16^n}.
$$
Esto se debe  a que lo que multiplica $1/16^k$ en cada término es menor que $16$ (en realidad $<1$ si $k>0$).

Para implementar la fórmula observemos  que
$$
\frac{1}{16^k} =  \frac{1}{2^{4k}} =\left(\frac{1}{2}\right)^{4k} = 0.5^{4k}
$$

In [ ]:
def calcular_pi_1997(n: int ) -> float:
    # ver: https://en.wikipedia.org/wiki/Approximations_of_%CF%80#Efficient_methods
    # calcula   \sum_{k=0}^n  0.5**(4*k) * (4/(8*k + 1) - 2/(8*k + 4) - 1/(8*k + 5) - 1/(8*k + 6))
    pi = 0
    for k in range(n+1):
        pi = pi +  0.5**(4*k) * (4/(8*k + 1) - 2/(8*k + 4) - 1/(8*k + 5) - 1/(8*k + 6))
    return pi

Pero si queremos implementarlo con mucha precisión lo mejor es usar el  módulo `decimal`. La escritura empeorará mucho, pero mejorará la precisión.

In [ ]:
from decimal import Decimal, getcontext
getcontext().prec = 500

def calcular_pi_1997(n: int ) -> float:
    # ver: https://en.wikipedia.org/wiki/Approximations_of_%CF%80#Efficient_methods
    # calcula   \sum_{k=0}^n  0.5**(4*k) * (4/(8*k + 1) - 2/(8*k + 4) - 1/(8*k + 5) - 1/(8*k + 6))
    pi = 0
    for k in range(n+1):
        pi = pi +  Decimal(0.5)**(4*k) * (Decimal(4)/Decimal(8*k + 1) - Decimal(2)/Decimal(8*k + 4) - Decimal(1)/Decimal(8*k + 5) - Decimal(1)/Decimal(8*k + 6))
    return pi

Si calculamos `calcular_pi_1997(n)` el error será menor que`Decimal(1) / Decimal(16**n)`. Probemos agunos casos:

In [ ]:
for n in [20, 30, 100, 200, 300]:
    print('Si n=', n)
    print(calcular_pi_1997(n))
    # print('tiene', str(-int(log10(Decimal(1) / Decimal(16**n)))), 'decimales correctos.\n')
    print('tiene error menor que ', str(Decimal(1) / Decimal(16**n)),'\n')

Si n= 20
3.141592653589793238462643381
tiene error menor que  8.271806125530276748714086921E-25 

Si n= 30
3.141592653589793238462643381
tiene error menor que  7.523163845262640050999913838E-37 

Si n= 100
3.141592653589793238462643381
tiene error menor que  3.872591914849318272818030633E-121 

Si n= 200
3.141592653589793238462643381
tiene error menor que  1.499696813895630954817644438E-241 

Si n= 300
3.141592653589793238462643381
tiene error menor que  5.807713756217503183283449999E-362 



Debemos decir que el cálculo de decimales  de $\pi$ no es de mucha utilidad y  actualmente se realizan ese tipo de cálculos más para probar la capacidad del hardware que para obtener los dígitos de $\pi$. Para hacer estos cálculos se utiliza la fórmula de Bailey-Borwein-Plouffe vista más  arriba  o  la fórmula  de Chudnovsky (https://es.wikipedia.org/wiki/Algoritmo_de_Chudnovsky).

### El número $e$

El número $e$ surge del estudio del interés compuesto, problema abordado por Jacob Bernoulli en 1683.

Si se invierte un \$1  con un interés del 100\% anual y
se pagan los intereses una vez al año, se obtendrán ${}$\$2.

Si se pagan los intereses 2 veces al año, dividiendo el interés entre 2, la cantidad obtenida es ${}$\$1 multiplicado por $1.5$ dos veces, es decir $1 \times 1.50^2 = 2.25$.

Si dividimos el año en 4 períodos (trimestres), al igual que la tasa de interés, se obtienen $1 \times 1.25^4 = 2.441\ldots$

En el caso de pagos mensuales el monto asciende a
$$
(1+\frac{1}{12})^{12} = 2.61303\ldots.
$$

Por tanto, si depositamos 1 peso en un banco que paga 100\% de interés y promete $n$ actualizaciones o capitalizaciones aun interes de 100/n\% cada una, el dinero que recuperaremos después de un año
\begin{equation}
\left(1 + \dfrac{1}{n}\right)^n \tag{2}
\end{equation}
pesos.

Finalmente, si vamos a un banco que ofrece 100% de interés con capitalización "instantanea",  contrario a lo que muchos puedan suponer, a lo largo de un año no vamos a entontranos con "infinito dinero" sinó con
\begin{equation}
e = \lim_{n \to \infty} \left(1 + \dfrac{1}{n}\right)^n = 2.71828\ldots \tag{3}
\end{equation}
pesos.

Una forma sencilla, en una computadora, de calcular $e$  es usar directamente la fórmula (2) como aproximación. Sin embargo,  esta fórmula introduce errores y no es buena computacionalmente.

Aqui  implementamos la fórmula (2)  y calculamos una aproximación de $e$ para 4N = 1000$.

In [ ]:
def calcular_e(n: int) -> float:
    return (1 + 1 / n)**n

e_aprox = calcular_e(1000)
print(e_aprox)

2.7169239322355936


La función `math.exp` eleva el número $e$ a la potencia indicada como argumento. Es decir `math.exp(1)` es una aproximación de Python al número $e$. Veamos cuanto difiere de la aproximación de $e$ calculada con la fórmula (2):

In [ ]:
import math
print(abs(e_aprox - math.exp(1)))

0.0013578962234515046


También se puede calcular $e$ usando la serie de  Taylor.
$$
e = \sum_{n=0}^\infty \frac{1}{n!}.
$$
Esta forma de calcular $e$ es más adecuada computacionalmente que la fórmula (2) y,  además, nos permite determinar fácilmente la precisión del resultado.   

**Ejercicio.** Dado $m$  entero, $m> 0$, implementar la aproximación por series de Taylor de $e$ en grado $m$:
$$
\sum_{n=0}^m \frac{1}{n!}.
$$
